tests/core/pyspec/eth2spec/utils/bls.py

```diff
@@ -121,24 +171,123 @@ def AggregatePKs(pubkeys):
-    if bls == py_ecc_bls:
-        return bls.SkToPk(SK)
+    if bls == py_ecc_bls or bls == arkworks_bls:  # no signature API in arkworks
+        return py_ecc_bls.SkToPk(SK)
     else:
         return bls.SkToPk(SK.to_bytes(32, 'big'))
 
 
 def pairing_check(values):
-    p_q_1, p_q_2 = values
-    final_exponentiation = final_exponentiate(
-        pairing(p_q_1[1], p_q_1[0], final_exponentiate=False)
-        * pairing(p_q_2[1], p_q_2[0], final_exponentiate=False)
-    )
-    return final_exponentiation == FQ12.one()
+    if bls == arkworks_bls or bls == fastest_bls:
+        p_q_1, p_q_2 = values
+        g1s = [p_q_1[0], p_q_2[0]]
+        g2s = [p_q_1[1], p_q_2[1]]
+        return arkworks_GT.multi_pairing(g1s, g2s) == arkworks_GT.one()
+    else:
+        p_q_1, p_q_2 = values
+        final_exponentiation = py_ecc_final_exponentiate(
+            py_ecc_pairing(p_q_1[1], p_q_1[0], final_exponentiate=False)
+            * py_ecc_pairing(p_q_2[1], p_q_2[0], final_exponentiate=False)
+        )
+        return final_exponentiation == py_ecc_GT.one()
+
+
+# Performs point addition of `lhs` and `rhs`
+# The points can either be in G1 or G2
+def add(lhs, rhs):
+    if bls == arkworks_bls or bls == fastest_bls:
+        return lhs + rhs
+    return py_ecc_add(lhs, rhs)
+
+
+# Performs Scalar multiplication between
+# `point` and `scalar`
+# `point` can either be in G1 or G2
+def multiply(point, scalar):
+    if bls == arkworks_bls or bls == fastest_bls:
+        int_as_bytes = scalar.to_bytes(32, 'little')
+        scalar = arkworks_Scalar.from_le_bytes(int_as_bytes)
+        return point * scalar
+    return py_ecc_mul(point, scalar)
+
+
+# Returns the point negation of `point`
+# `point` can either be in G1 or G2
+def neg(point):
+    if bls == arkworks_bls or bls == fastest_bls:
+        return -point
+    return py_ecc_neg(point)
+
+
+# Returns the identity point in G1
+def Z1():
+    if bls == arkworks_bls or bls == fastest_bls:
+        return arkworks_G1.identity()
+    return py_ecc_Z1
+
+
+# Returns the chosen generator point in G1
+def G1():
+    if bls == arkworks_bls or bls == fastest_bls:
+        return arkworks_G1()
+    return py_ecc_G1
+
+
+# Returns the chosen generator point in G2
+def G2():
+    if bls == arkworks_bls or bls == fastest_bls:
+        return arkworks_G2()
+    return py_ecc_G2
+
+
+# Serializes a point in G1
+# Returns a bytearray of size 48 as
+# we use the compressed format
+def G1_to_bytes48(point):
+    if bls == arkworks_bls or bls == fastest_bls:
+        return point.to_compressed_bytes()
+    return py_ecc_G1_to_bytes48(point)
+
+
+# Serializes a point in G2
+# Returns a bytearray of size 96 as
+# we use the compressed format
+def G2_to_bytes96(point):
+    if bls == arkworks_bls or bls == fastest_bls:
+        return point.to_compressed_bytes()
+    return py_ecc_G2_to_bytes96(point)
+
+
+# Deserializes a purported compressed serialized
+# point in G1
+# - No subgroup checks are performed
```

Comment:
Why are the subgroup checks not performed? Are they too slow?

(For reference purpose) `from_compressed_bytes` does the subgroup check which then triggers https://docs.rs/ark-bls12-381/latest/src/ark_bls12_381/curves/g1.rs.html#95

Reply:
Yeah the arkworks library exposes a method to also do subgroup checks -- It's not done because I checked py-ecc and they were not doing it either. In the 4844 specs, the subgroup checks are instead done in `KeyValidate`.

I guess this way makes them more explicit in the specs and there are times when you are deserializing trusted input so you know that it is in the correct subgroup already.

We _could_ do subgroup checks whenever we deserialise iff arkworks is fast enough to still generate the tests, but I'd prefer to not change the semantics of those methods in this PR tbh